File: articles/developmental-systems-theory-and-the-ecology-of-human-growth/go/developmental_systems_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func col(header []string, name string) int {
	for i, h := range header {
		if h == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run developmental_systems_summary.go data/developmental_systems_panel.csv")
	}
	f, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	rows, err := csv.NewReader(f).ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	developmentIdx := col(header, "development_score")
	supportIdx := col(header, "ecological_support")
	stressIdx := col(header, "ecological_stress")
	familyIdx := col(header, "current_family")
	peerIdx := col(header, "current_peer")
	if developmentIdx < 0 || supportIdx < 0 || stressIdx < 0 || familyIdx < 0 || peerIdx < 0 {
		log.Fatal("Required columns missing")
	}

	count := 0
	developmentSum, supportSum, stressSum, familySum, peerSum := 0.0, 0.0, 0.0, 0.0, 0.0

	for _, row := range rows[1:] {
		development, e1 := strconv.ParseFloat(row[developmentIdx], 64)
		support, e2 := strconv.ParseFloat(row[supportIdx], 64)
		stress, e3 := strconv.ParseFloat(row[stressIdx], 64)
		family, e4 := strconv.ParseFloat(row[familyIdx], 64)
		peer, e5 := strconv.ParseFloat(row[peerIdx], 64)
		if e1 != nil || e2 != nil || e3 != nil || e4 != nil || e5 != nil {
			continue
		}
		count++
		developmentSum += development
		supportSum += support
		stressSum += stress
		familySum += family
		peerSum += peer
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean development_score: %.4f\n", developmentSum/float64(count))
	fmt.Printf("Mean ecological_support: %.4f\n", supportSum/float64(count))
	fmt.Printf("Mean ecological_stress: %.4f\n", stressSum/float64(count))
	fmt.Printf("Mean current_family: %.4f\n", familySum/float64(count))
	fmt.Printf("Mean current_peer: %.4f\n", peerSum/float64(count))
}

File: articles/developmental-systems-theory-and-the-ecology-of-human-growth/c/developmental_systems_summary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE 16384

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s data/developmental_systems_panel.csv\n", argv[0]);
        return 1;
    }

    FILE *fp = fopen(argv[1], "r");
    if (!fp) {
        perror("Unable to open CSV file");
        return 1;
    }

    char line[MAX_LINE];
    if (!fgets(line, sizeof(line), fp)) {
        fprintf(stderr, "CSV is empty\n");
        fclose(fp);
        return 1;
    }

    int development_col = -1, support_col = -1, stress_col = -1, family_col = -1, peer_col = -1, col = 0;
    char *token = strtok(line, ",");

    while (token != NULL) {
        token[strcspn(token, "\r\n")] = 0;
        if (strcmp(token, "development_score") == 0) development_col = col;
        if (strcmp(token, "ecological_support") == 0) support_col = col;
        if (strcmp(token, "ecological_stress") == 0) stress_col = col;
        if (strcmp(token, "current_family") == 0) family_col = col;
        if (strcmp(token, "current_peer") == 0) peer_col = col;
        token = strtok(NULL, ",");
        col++;
    }

    if (development_col < 0 || support_col < 0 || stress_col < 0 || family_col < 0 || peer_col < 0) {
        fprintf(stderr, "Required columns not found\n");
        fclose(fp);
        return 1;
    }

    long count = 0;
    double development_sum = 0.0, support_sum = 0.0, stress_sum = 0.0, family_sum = 0.0, peer_sum = 0.0;

    while (fgets(line, sizeof(line), fp)) {
        int current_col = 0;
        double development = 0.0, support = 0.0, stress = 0.0, family = 0.0, peer = 0.0;
        char *field = strtok(line, ",");

        while (field != NULL) {
            if (current_col == development_col) development = atof(field);
            if (current_col == support_col) support = atof(field);
            if (current_col == stress_col) stress = atof(field);
            if (current_col == family_col) family = atof(field);
            if (current_col == peer_col) peer = atof(field);
            field = strtok(NULL, ",");
            current_col++;
        }

        count++;
        development_sum += development;
        support_sum += support;
        stress_sum += stress;
        family_sum += family;
        peer_sum += peer;
    }

    fclose(fp);

    printf("Rows analyzed: %ld\n", count);
    printf("Mean development_score: %.4f\n", development_sum / count);
    printf("Mean ecological_support: %.4f\n", support_sum / count);
    printf("Mean ecological_stress: %.4f\n", stress_sum / count);
    printf("Mean current_family: %.4f\n", family_sum / count);
    printf("Mean current_peer: %.4f\n", peer_sum / count);
    return 0;
}
